contexts/PaymentContext/tests/Unit/Domain/LessSimpleTransferCodeGeneratorTest.php

```diff
@@ -0,0 +1,81 @@
+<?php
+
+declare( strict_types = 1 );
+
+namespace WMDE\Fundraising\Frontend\PaymentContext\Tests\Unit\Domain;
+
+use PHPUnit\Framework\TestCase;
+use WMDE\Fundraising\Frontend\PaymentContext\Domain\LessSimpleTransferCodeGenerator;
+
+/**
+ * @covers \WMDE\Fundraising\Frontend\PaymentContext\Domain\LessSimpleTransferCodeGenerator
+ *
+ * @licence GNU GPL v2+
+ */
+class LessSimpleTransferCodeGeneratorTest extends TestCase {
+
+	/**
+	 * @dataProvider characterAndCodeProvider
+	 */
+	public function testGenerateBankTransferCode( string $expectedCode, string $usedCharacters ): void {
+		$generator = LessSimpleTransferCodeGenerator::newDeterministicGenerator(
+			$this->newFixedCharacterGenerator( $usedCharacters )
+		);
+
+		$this->assertSame( $expectedCode, $generator->generateTransferCode() );
+	}
+
+	public function characterAndCodeProvider(): iterable {
+		yield [ 'ACD-EFK-X', 'ACDEFKLMNPRSTWXYZ349ACDEF' ];
+		yield [ 'AAA-AAA-9', 'AAAAAAAAAAAAAAAAAAAAAAAAA' ];
+		yield [ 'CAA-AAA-A', 'CAAAAAAAAAAAAAAAAAAAAAAAA' ];
+		yield [ 'ACA-CAC-4', 'ACACACACACACACACACACACACA' ];
+	}
+
+	private function newFixedCharacterGenerator( string $characters ): \Generator {
+		yield from str_split( $characters );
+	}
+
+	/**
+	 * @dataProvider invalidCodeProvider
+	 */
+	public function testInvalidTransferCodesAreNotValid( string $invalidCode ): void {
+		$this->assertFalse( $this->newGenerator()->transferCodeIsValid( $invalidCode ) );
+	}
+
+	private function newGenerator(): LessSimpleTransferCodeGenerator {
+		return LessSimpleTransferCodeGenerator::newDeterministicGenerator(
+			$this->newFixedCharacterGenerator( 'ACDEFKLMNPRSTWXYZ349' )
+		);
+	}
+
+	public function invalidCodeProvider(): iterable {
+		yield 'Empty code' => [ '' ];
+		yield 'Without checksum' => [ 'ACD-EFK-' ];
+		yield 'Missing dash' => [ 'ACDEFK-X' ];
+		yield 'Missing checksum dash' => [ 'ACD-EFK' ];
+		yield 'Missing both dashes' => [ 'ACDEFK' ];
+		yield 'Extra dash' => [ 'ACD-EFK-X-' ];
```

Comment:
These should now be valid no? @gbirke

Reply:
I thought about it when I read them and was unsure. Because we're not the not the ones validating the codes (that's done by the Gruen software), it's more of a courtesy or "reference implementation" of what we think valid codes should be. But then we'd also need to incorporate the 2-char prefixes in the tests. Yesterday in the story time we defined that they should be `XW` (for personalized donations) and `XR` (for anonymous donations). And at the beginning of this issue you wrote  

> Did not include the prefix in this class as it is not clear to me if
it should go in there or not. If it depends on something else, such
as the type of donation, then probably not.

As it turns out, it depends on the type of data in the donation. I would have included the fixed prefix in the `TransferCode` class constructor (simulating an enum type with constants of `TransferCode` and checks inside the constructor), but I don't feel strongly about it.

Questions to @KaiNissen : 

1) Should we implement `validateTransferCode` at all in the Frontend or just use the code as an example in the documentation for Gruen?
2) If we leave the code in, does it make sense to include the prefix in the `LessSimpleTransferCodeGenerator`? At least it must be part of the checks in  `validateTransferCode` because it must be passed to the checksum calculation.